src/LuaTransformer.ts

```diff
@@ -5113,4 +5118,37 @@ export class LuaTransformer {
             return visitResult;
         }
     }
+
+    public generateConstructorDecorationStatement(
+        declaration: ts.ClassLikeDeclaration
+    ): tstl.AssignmentStatement | undefined {
+        const className = declaration.name !== undefined
+            ? this.transformIdentifier(declaration.name)
+            : tstl.createAnonymousIdentifier();
+
+        const decorators = declaration.decorators;
+        if (!decorators) { return undefined; }
+
+        const decoratorExpressions: tstl.Expression[] = [];
+        for (const decorator of decorators) {
+            const expressionVisitResult = this.transformExpression(decorator.expression);
+            if (expressionVisitResult !== undefined) {
+                decoratorExpressions.push(expressionVisitResult);
+            }
+        }
+
+        const decoratorArguments: tstl.Expression[] = [];
```

Comment:
Since we can't have a helper accept both functions with and without context I think we should add context there and add a check for function kind.

Reply:
Can you elaborate on this for me? I'm not really sure what this means.

Reply:
Ah, don't mind this line reference, I was a bit confused because all decorator factories in tests have `this: void` (they probably shouldn't?).

Inner decorator function still should be checked for function kind. If it has `this: void` it would transpile successfully, but fail at runtime. See `validateFunctionAssignment` for a similar check used when function is passed as a parameter.

Reply:
For clarification: The (precompiled) Lualib function always calls the decorators with context. If a callback is passed with `this: void` it should error out.